lumicks/pylake/kymotracker/detail/gaussian_mle.py

```diff
@@ -0,0 +1,138 @@
+    )
+    jac_fcn = partial(
+        poisson_log_likelihood_jacobian,
+        expectation_fcn=expectation_fcn,
+        derivatives_fcn=derivatives_fcn,
+        photon_count=photon_count,
+    )
+
+    return minimize(
+        optim_fcn,
+        initial_guess,
+        jac=jac_fcn,
+        method="L-BFGS-B",
+        bounds=bounds,
+    )
+
+
+def normal_pdf_1d(x, center, width):
+    """Evaluates the probablitity density function for a 1D normal distribution
+
+    Parameters
+    ----------
+    x : array-like
+        Position data at which the function is to be evaluated.
+    center : float
+        Distribution center.
+    width : float
+        Distribution width; square root of the variance.
+    """
+    norm_factor = 1 / np.sqrt(2 * np.pi * width ** 2)
+    return norm_factor * np.exp(-0.5 * ((x - center) / width) ** 2)
+
+
+def peak_expectation_1d(x, total_photons, center, width, background, pixel_size):
+    """Calculates the expectation value of a gaussian peak evaluated at x with baseline offset.
+
+    Parameters
+    ----------
+    x : array-like
+        Position data at which the function is to be evaluated.
+    total_photons : float
+        Total number of photons emitted by the imaged particle.
+    center : float
+        Peak center in nm.
+    width : float
+        Peak width in nm.
+    background : float
+        Square root of the photonic background per pixel.
+    pixel_size : float
+        Pixel size in nm.
+    """
+    return total_photons * pixel_size * normal_pdf_1d(x, center, width) + background ** 2
+
+
+def peak_expectation_1d_derivatives(x, total_photons, center, width, background, pixel_size):
+    """Evaluate the derivatives of the expectation w.r.t. each parameter."""
+    pdf = normal_pdf_1d(x, center, width)
+    d_dphotons = pixel_size * pdf
+    d_dcenter = total_photons * pixel_size / width ** 2 * pdf * (x - center)
+    d_dwidth = total_photons * pixel_size * pdf * ((x - center) ** 2 - width ** 2) / width ** 3
+    d_dbackground = 2 * background
+    return d_dphotons, d_dcenter, d_dwidth, d_dbackground
+
+
+def gaussian_mle_1d(x, photon_count, pixel_size):
+    """Calculate the maximum likelihood estimate of the model parameters given measured photon count for 1D data.
+
+    Parameters
+    ----------
+    x : array-like
+        Position data at which the function is to be evaluated.
+    photon_count : array-like
+        Measured photon counts at each position.
+    pixel_size : float
+        Pixel size in nm
+    """
+    expectation_fcn = partial(peak_expectation_1d, x, pixel_size=pixel_size)
+    derivatives_fcn = partial(peak_expectation_1d_derivatives, x, pixel_size=pixel_size)
+    initial_guess = (
+        np.max(photon_count) / pixel_size * np.sqrt(2 * np.pi * 200 ** 2),
+        x[np.argmax(photon_count)],
+        200,
+        1,
+    )
+    bounds = ((0.01, None), (np.min(x), np.max(x)), (100, 1000), (np.finfo(float).eps, None))
```

Comment:
Weird. I could have sworn that I commented on this line already.
The minimum offset, are we sure that a lower bound of machine epsilon is enough to not result in numerical problems?

The range in sigma seems a bit arbitrary? Are we certain nothing would ever be outside those bounds?

Reply:
I changed the bounds for `sigma` to be `pixel_size` for the lower bound (anything with a width lower than the pixel size would just be a spike, which would cause problems for fitting anyway) and for the upper bound `10 * pixel_size` which I think is a reasonable limit. Here we don't want to let the `sigma` blow up so much that the amplitude drops to zero

Reply:
👍